src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.DesignerAttribute;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal class RemoteDesignerAttributeIncrementalAnalyzer : IncrementalAnalyzerBase
+    {
+        private const string DataKey = "DesignerAttributeData";
+
+        private readonly Workspace _workspace;
+
+        /// <summary>
+        /// Channel back to VS to inform it of the designer attributes we discover.
+        /// </summary>
+        private readonly RemoteEndPoint _endPoint;
+
+        /// <summary>
+        /// Storage where we can keep track of what we know about the project and have informed VS
+        /// (and the project systems) about.
+        /// </summary>
+        private readonly IPersistentStorage _storage;
+
+        public RemoteDesignerAttributeIncrementalAnalyzer(Workspace workspace, RemoteEndPoint endPoint)
+        {
+            _workspace = workspace;
+            _endPoint = endPoint;
+
+            var storageService = _workspace.Services.GetRequiredService<IPersistentStorageService>();
+            _storage = storageService.GetStorage(workspace.CurrentSolution);
+        }
+
+        public override Task AnalyzeProjectAsync(Project project, bool semanticsChanged, InvocationReasons reasons, CancellationToken cancellationToken)
+            => AnalyzeProjectAsync(project, specificDoc: null, cancellationToken);
+
+        public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, InvocationReasons reasons, CancellationToken cancellationToken)
```

Comment:
>SyntaxNode [](start = 69, length = 10)

Shouldn't this be `SyntaxNode?` (also, w/o `Opt` suffix).

Reply:
sure.

Reply:
(looks like base type isn't NRTified)

Reply:
Done.